File: typescript/src/screen.ts
#!/usr/bin/env ts-node
import { execSync } from "child_process";
import * as path from "path";
import * as dotenv from "dotenv";

dotenv.config();

// ─── Types ───────────────────────────────────────────────────────────────────

interface DateObj {
  year?: number;
  month?: number;
  day?: number;
}

interface Experience {
  companyName?: string;
  title?: string;
  timePeriod?: { startDate?: DateObj; endDate?: DateObj };
  company?: { employeeCountRange?: { start?: number; end?: number }; industries?: string[] };
  locationName?: string;
}

interface Education {
  schoolName?: string;
  degreeName?: string;
  fieldOfStudy?: string;
  timePeriod?: { startDate?: DateObj; endDate?: DateObj };
}

interface ProfileData {
  firstName?: string;
  lastName?: string;
  headline?: string;
  locationName?: string;
  experience?: Experience[];
  education?: Education[];
}

interface FetchResult {
  profile?: ProfileData;
  contact?: Record<string, unknown>;
  error?: string;
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

const divider = (char = "─", width = 52) => console.log(char.repeat(width));
const section = (title: string) => {
  console.log(`\n  ${title.toUpperCase()}`);
  console.log("  " + "─".repeat(title.length + 2));
};

function formatYear(date?: DateObj): string | null {
  return date?.year ? String(date.year) : null;
}

function companyAge(founded?: number): string {
  if (!founded) return "—";
  return `${new Date().getFullYear() - founded} years`;
}

// ─── Fetch data via Python subprocess ────────────────────────────────────────

function fetchProfile(url: string): FetchResult {
  const script = path.join(__dirname, "fetch_profile.py");
  try {
    const output = execSync(`python3 "${script}" "${url}"`, {
      env: process.env,
      timeout: 30000,
    }).toString();
    return JSON.parse(output);
  } catch (err: any) {
    const msg = err.stdout?.toString() || err.message || "Unknown error";
    try {
      return JSON.parse(msg);
    } catch {
      return { error: msg };
    }
  }
}

// ─── Crunchbase lookup (optional) ────────────────────────────────────────────

function getFoundedCrunchbase(companyName: string): number | null {
  const key = process.env.CRUNCHBASE_API_KEY;
  if (!key || !companyName) return null;
  try {
    const body = JSON.stringify({
      field_ids: ["name", "founded_on"],
      query: [{ type: "predicate", field_id: "name", operator_id: "contains", values: [companyName] }],
      limit: 1,
    });
    const out = execSync(
      `curl -s -X POST "https://api.crunchbase.com/api/v4/searches/organizations" \
       -H "X-cb-user-key: ${key}" -H "Content-Type: application/json" \
       -d '${body.replace(/'/g, "'\\''")}'`,
      { timeout: 10000 }
    ).toString();
    const data = JSON.parse(out);
    const founded = data?.entities?.[0]?.properties?.founded_on;
    if (typeof founded === "string") return parseInt(founded.slice(0, 4));
    if (typeof founded === "object") return parseInt(String(founded?.value).slice(0, 4));
  } catch { /* optional */ }
  return null;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  const url = process.argv[2]?.trim();

  if (!url) {
    console.log("Usage: npx ts-node src/screen.ts <linkedin_profile_url>");
    process.exit(1);
  }

  if (!process.env.LINKEDIN_USER || !process.env.LINKEDIN_PASS) {
    console.error("❌  LINKEDIN_USER and LINKEDIN_PASS must be set.");
    console.error("    Create a .env file or export them before running.");
    process.exit(1);
  }

  console.log();
  divider("═");
  console.log("  🏠  HACKERHOUSE SCREENER");
  divider("═");
  console.log(`  Fetching: ${url}`);
  divider();

  const result = fetchProfile(url);

  if (result.error) {
    console.error(`❌  ${result.error}`);
    process.exit(1);
  }

  const p = result.profile!;
  const name = [p.firstName, p.lastName].filter(Boolean).join(" ") || "Unknown";
  const headline = p.headline || "";
  const location = p.locationName || "";

  console.log(`\n  👤  ${name}`);
  if (headline) console.log(`      ${headline}`);
  if (location) console.log(`      📍 ${location}`);

  // ── Current company ──────────────────────────────────────────────
  const experiences = p.experience || [];
  const current = experiences.find((e) => !e.timePeriod?.endDate);

  let companyFounded: number | null = null;

  if (current) {
    const name_ = current.companyName || "—";
    const title = current.title || "—";
    const startY = formatYear(current.timePeriod?.startDate) || "";
    const industries = current.company?.industries?.join(", ") || "—";
    const sizeRange = current.company?.employeeCountRange;
    const size = sizeRange ? `${sizeRange.start}–${sizeRange.end}` : "—";

    // Try Crunchbase for founding year
    companyFounded = getFoundedCrunchbase(name_);

    section("Current Company");
    console.log(`  Company   : ${name_}`);
    console.log(`  Title     : ${title}`);
    console.log(`  Founded   : ${companyFounded ?? "—"}`);
    console.log(`  Age       : ${companyAge(companyFounded ?? undefined)}`);
    console.log(`  Size      : ${size}`);
    console.log(`  Industry  : ${industries}`);
  } else {
    section("Current Company");
    console.log("  No current position found.");
  }

  // ── Experience ───────────────────────────────────────────────────
  section("Experience");
  if (experiences.length) {
    for (const exp of experiences) {
      const co = exp.companyName || "—";
      const title = exp.title || "—";
      const startY = formatYear(exp.timePeriod?.startDate);
      const endDate = exp.timePeriod?.endDate;
      const endY = endDate ? formatYear(endDate) : "present";
      const years = startY ? `${startY} – ${endY}` : (endY || "");
      const tag = !exp.timePeriod?.endDate ? " ← current" : "";
      console.log(`  • ${title} @ ${co}  [${years}]${tag}`);
    }
  } else {
    console.log("  No experience listed.");
  }

  // ── Education ────────────────────────────────────────────────────
  section("Education");
  const education = p.education || [];
  if (education.length) {
    for (const edu of education) {
      const school = edu.schoolName || "—";
      const deg = [edu.degreeName, edu.fieldOfStudy].filter(Boolean).join(", ") || "—";
      const startY = formatYear(edu.timePeriod?.startDate);
      const endY = formatYear(edu.timePeriod?.endDate);
      const years = startY ? `${startY} – ${endY || ""}` : "";
      console.log(`  • ${school}`);
      console.log(`    ${deg}${years ? "  (" + years + ")" : ""}`);
    }
  } else {
    console.log("  No education listed.");
  }

  console.log();
  divider("═");
  console.log();
}

main().catch((e) => {
  console.error("❌ ", e.message);
  process.exit(1);
});

File: screen.py
#!/usr/bin/env python3
"""
Hackerhouse LinkedIn Screener
Uses: RapidAPI - Fresh LinkedIn Profile Data
Usage: python screen.py <linkedin_url>
"""

import sys
import os
import requests
from datetime import datetime

RAPIDAPI_KEY = os.environ.get("RAPIDAPI_KEY", "")
CRUNCHBASE_API_KEY = os.environ.get("CRUNCHBASE_API_KEY", "")

RAPIDAPI_HOST = "fresh-linkedin-profile-data.p.rapidapi.com"


def get_linkedin_profile(url):
    resp = requests.get(
        "https://fresh-linkedin-profile-data.p.rapidapi.com/get-linkedin-profile",
        headers={
            "X-RapidAPI-Key": RAPIDAPI_KEY,
            "X-RapidAPI-Host": RAPIDAPI_HOST,
        },
        params={"linkedin_url": url, "include_skills": "false"},
    )
    resp.raise_for_status()
    result = resp.json()
    return result.get("data", result)


def get_company_profile(company_linkedin_url):
    resp = requests.get(
        "https://fresh-linkedin-profile-data.p.rapidapi.com/get-company-by-linkedinurl",
        headers={
            "X-RapidAPI-Key": RAPIDAPI_KEY,
            "X-RapidAPI-Host": RAPIDAPI_HOST,
        },
        params={"linkedin_url": company_linkedin_url},
    )
    if resp.status_code == 200:
        result = resp.json()
        return result.get("data", result)
    return None


def get_founded_crunchbase(company_name):
    if not CRUNCHBASE_API_KEY or not company_name:
        return None
    resp = requests.post(
        "https://api.crunchbase.com/api/v4/searches/organizations",
        headers={"X-cb-user-key": CRUNCHBASE_API_KEY, "Content-Type": "application/json"},
        json={
            "field_ids": ["name", "founded_on"],
            "query": [{"type": "predicate", "field_id": "name", "operator_id": "contains", "values": [company_name]}],
            "limit": 1,
        },
    )
    if resp.status_code == 200:
        entities = resp.json().get("entities", [])
        if entities:
            founded = entities[0].get("properties", {}).get("founded_on", {})
            return founded.get("value") if isinstance(founded, dict) else founded
    return None


def divider(char="─", width=50):
    print(char * width)


def section(title):
    print(f"\n  {title.upper()}")
    print("  " + "─" * (len(title) + 2))


def parse_year(date_obj):
    if not date_obj:
        return None
    if isinstance(date_obj, dict):
        return date_obj.get("year")
    if isinstance(date_obj, str) and len(date_obj) >= 4:
        return date_obj[:4]
    return None


def main():
    if not RAPIDAPI_KEY:
        print("❌  RAPIDAPI_KEY not set.")
        print("    Export it: export RAPIDAPI_KEY=your_key")
        sys.exit(1)

    if len(sys.argv) < 2:
        print("Usage: python screen.py <linkedin_profile_url>")
        sys.exit(1)

    url = sys.argv[1].strip()

    print()
    divider("═")
    print("  🏠  HACKERHOUSE SCREENER")
    divider("═")
    print(f"  Fetching: {url}")
    divider()

    try:
        profile = get_linkedin_profile(url)
    except requests.HTTPError as e:
        print(f"❌  RapidAPI error {e.response.status_code}: {e.response.text}")
        sys.exit(1)
    except Exception as e:
        print(f"❌  {e}")
        sys.exit(1)

    # Basic info
    name = profile.get("full_name") or "Unknown"
    headline = profile.get("headline") or ""
    location = profile.get("city") or profile.get("location") or ""
    print(f"\n  👤  {name}")
    if headline:
        print(f"      {headline}")
    if location:
        print(f"      📍 {location}")

    # --- Current company ---
    experiences = profile.get("experiences") or []
    current = next((e for e in experiences if not e.get("ends_at") and not e.get("end_year")), None)

    current_company_name = None
    company_founded = None
    company_age = None
    company_size = None
    company_industry = None

    if current:
        current_company_name = current.get("company") or current.get("company_name")
        company_linkedin_url = current.get("company_linkedin_url") or current.get("company_linkedin_profile_url")

        if company_linkedin_url:
            co = get_company_profile(company_linkedin_url)
            if co:
                company_founded = co.get("founded_year") or co.get("founded")
                company_size = co.get("company_size") or co.get("company_size_on_linkedin")
                company_industry = co.get("industry")

        if not company_founded and current_company_name:
            company_founded = get_founded_crunchbase(current_company_name)

        if company_founded:
            try:
                company_age = datetime.now().year - int(str(company_founded)[:4])
            except:
                pass

    section("Current Company")
    if current_company_name:
        print(f"  Company   : {current_company_name}")
        print(f"  Title     : {current.get('title') or '—'}")
        print(f"  Founded   : {company_founded or '—'}")
        print(f"  Age       : {str(company_age) + ' years' if company_age is not None else '—'}")
        print(f"  Size      : {company_size or '—'}")
        print(f"  Industry  : {company_industry or '—'}")
    else:
        print("  No current company found.")

    # --- All experience ---
    section("Experience")
    if experiences:
        for exp in experiences:
            company = exp.get("company") or exp.get("company_name") or "—"
            title = exp.get("title") or exp.get("position") or "—"
            start_y = exp.get("start_year") or parse_year(exp.get("starts_at"))
            ends_at = exp.get("ends_at") or exp.get("end_year")
            is_current = not ends_at
            end_y = "present" if is_current else (exp.get("end_year") or parse_year(exp.get("ends_at")) or "")
            years = f"{start_y} – {end_y}" if start_y else (str(end_y) if end_y else "")
            tag = " ← current" if is_current else ""
            print(f"  • {title} @ {company}  [{years}]{tag}")
    else:
        print("  No experience listed.")

    # --- Education ---
    section("Education")
    education = profile.get("educations") or profile.get("education") or []
    if education:
        for edu in education:
            school = edu.get("school") or edu.get("institution") or "—"
            degree = edu.get("degree") or edu.get("degree_name") or ""
            field = edu.get("field_of_study") or edu.get("field") or ""
            deg_str = ", ".join(filter(None, [degree, field])) or "—"
            start_y = edu.get("start_year") or parse_year(edu.get("starts_at"))
            end_y = edu.get("end_year") or parse_year(edu.get("ends_at"))
            years = f"{start_y} – {end_y}" if start_y else ""
            print(f"  • {school}")
            print(f"    {deg_str}  {('(' + str(years) + ')') if years else ''}")
    else:
        print("  No education listed.")

    print()
    divider("═")
    print()


if __name__ == "__main__":
    main()
